akka-stream-tests/src/test/scala/akka/stream/io/InputStreamSinkSpec.scala

```diff
@@ -0,0 +1,108 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.io
+
+import java.io.IOException
+import java.util.concurrent.TimeoutException
+
+import akka.stream.scaladsl.Keep
+import akka.stream.testkit.AkkaSpec
+import akka.stream.testkit.Utils._
+import akka.stream.testkit.scaladsl.TestSource
+import akka.stream.{ ActorMaterializer, ActorMaterializerSettings }
+import akka.util.ByteString
+import org.scalatest.concurrent.ScalaFutures
+
+import scala.concurrent.duration._
+import scala.concurrent.{ Await, Future }
+
+class InputStreamSinkSpec extends AkkaSpec(UnboundedMailboxConfig) with ScalaFutures {
+
+  val settings = ActorMaterializerSettings(system).withDispatcher("akka.actor.default-dispatcher")
+  implicit val materializer = ActorMaterializer(settings)
+  val timeout = 300.milliseconds
+  val byteString = ByteString("abc")
+
+  def expectBlocked[T](f: Future[T]) =
+    the[Exception] thrownBy Await.result(f, timeout) shouldBe a[TimeoutException]
+
+  def expectSuccess[T](f: Future[T], value: T) =
+    Await.result(f, timeout) should be(value)
+
+  "InputStreamSink" must {
+    "read bytes from InputStream" in assertAllStagesStopped {
+      val (probe, (inputStream, f)) = TestSource.probe[ByteString].toMat(InputStreamSink())(Keep.both).run()
+      probe.sendNext(byteString)
+      val arr = new Array[Byte](3)
+      inputStream.read(arr)
+      assert(arr === "abc".toArray)
+
+      probe.sendComplete()
+      expectSuccess(f, 3)
+    }
+
+    "read call blocks until get from upstream" in assertAllStagesStopped {
+      import scala.concurrent.ExecutionContext.Implicits.global
+      val (probe, (inputStream, _)) = TestSource.probe[ByteString].toMat(InputStreamSink())(Keep.both).run()
+
+      val arr = new Array[Byte](3)
+
+      val f = Future(inputStream.read(arr))
+
+      expectBlocked(f)
+      probe.sendNext(byteString)
+      expectSuccess(f, 3)
+
+      probe.sendComplete()
+    }
+
+    "filling up buffer by default" in assertAllStagesStopped {
+      import scala.concurrent.ExecutionContext.Implicits.global
+      val (probe, (inputStream, _)) = TestSource.probe[ByteString].toMat(InputStreamSink())(Keep.both).run()
+
+      probe.sendNext(ByteString("abc"))
+      probe.sendNext(ByteString("def"))
+
+      val arr1 = new Array[Byte](3)
+      val arr2 = new Array[Byte](3)
+      val f1 = Future(inputStream.read(arr1))
+      val f2 = Future(inputStream.read(arr2))
+      Await.result(f1, timeout) should be(3)
+      Await.result(f2, timeout) should be(3)
+
+      assert(arr1 === "abc".toArray)
+      assert(arr2 === "def".toArray)
+
+      probe.sendComplete()
+    }
```

Comment:
last read if it's in blocked state will return everything that is in buffer (assuming less then client requested). Then IO exception will be thrown. Will add this test if it's not yet there